Add AML-Dashboard (#1)
eProsima-Private/AML-Dashboard: README.md

# AML-Dashboard

## Installation

To get started with AML-Dashboard, follow these steps.

Clone this repository using the following command:

```bash
git clone https://github.com/eProsima-Private/AML-Dashboard.git
```

### Building from sources

If **AML-IP** is not already installed in your system, follow one of the methods outlined in its [documentation site](https://aml-ip.readthedocs.io/en/latest/rst/developer_manual/installation/sources/linux/linux.html).

> ### Requirements

Reviewer: Add file requirements.txt same as dockerfile:  pip install -r requirements.txt

 pip3 install -U colcon-common-extensions \
        vcstool \
        astropy==5.3.4 \
        blinker==1.7.0 \
        cffi==1.16.0 \
        click==8.1.7 \
        Flask==3.0.0 \
        Flask-Cors==4.0.0 \
        itsdangerous==2.1.2 \
        Jinja2==3.1.2 \
        joblib==1.3.2 \
        MarkupSafe==2.1.3 \
        msgpack==1.0.7 \
        numpy==1.26.1 \
        packaging==23.2 \
        pandas==2.1.2 \
        pycparser==2.21 \
        pyerfa==2.0.1.1 \
        python-dateutil==2.8.2 \
        pytz==2023.3.post1 \
        PyYAML==6.0.1 \
        scikit-learn==1.3.2 \
        scipy==1.11.3 \
        six==1.16.0 \
        threadpoolctl==3.2.0 \
        tqdm==4.66.1 \
        tzdata==2023.3 \
        Werkzeug==3.0.1

Author: I agree with adding the `requirements.txt` file. However, please ensure that only the dependencies specific to the dashboard are included in this file. Dependencies for AML-IP should be listed in the AML-IP documentation to keep them separate.

> 

Reviewer: Add that this version of aml_engine works with python 3.11

> Before proceeding further, ensure that you have **npm** (8.5.1) and **nodejs** (v12.22.9) installed on your system. If not, install them using the following commands:

Reviewer: Make sure that context broker is correctly installed. Link to installation of context broker.

```bash
http://localhost:5173/
```

### Data Management Tab

Efficiently manage and create new datasets.

To create a dataset, simply follow these steps:

1. Activate the `video` switch in the *webcam* section to enable webcam capture.
2. Specify the label of the dataset in the *Instance label* section.
3. Click on the `Hold to record instances` button in the *Capture instances to the training set* to start recording instances.
4. Once recorded, the dataset will be promptly displayed in the *dataset browser* section for easy access and management.

<div align="center">
  <img src=".figures/aml-dashboard_data_management.png" width="750">
</div>

### Training Tab

Effortlessly train models using either AML or Neural Network.

To train a model using AML, follow these steps:

1. Specify the number of parallel trainings (executions) you wish to run.
2. Define the number of iterations per execution.
3. Set the percentage of dataset to distribute in each execution.
4. Click on the `Train` button in the *AML Training Launcher* to initiate the training process.
5. Once the training is completed, the model status will appear as **Finished :)** in the *AML Status*.

> * Please ensure that at least one **Computing Node** is running to facilitate the training process.

Reviewer: Put before step 1.

<div align="center">
  <img src=".figures/aml-dashboard_training.png" width="750">
</div>

### Fetching Tab

Retrieve trained AML Models effortlessly.

To fetch a model, follow these steps:

1. Click on the `Search for statistics` button in the *AML Statistics Fetcher*.
2. Once the statistics are received, the status will appear as **Statistics received !** in the *AML Collaborative Learning Status*.
3. Click on the `Request model` button in the *AML Model Fetcher*.
4. Once the model is received, the status will change to **Model received !** in the *AML Collaborative Learning Status*.

* Make sure that at least one **Model Manager Sender Node** is running to facilitate the model fetching process.

> 

Reviewer: Put before step 1.

<div align="center">
  <img src=".figures/aml-dashboard_fetching.png" width="750">
</div>

### Batch Prediction Tab

Effortlessly predict the output of a batch of inputs using an AML Model.

To predict the output of a dataset, follow these steps:

1. Click on the `Update predictions` button in the *Algebraic Machine Learning*.
2. The predictions will be displayed in the *Results Algebraic Machine Learning* plot.

> * Ensure that you have access to at least one AML Model to facilitate the batch prediction process.

Reviewer: Put before step 1.

<div align="center">
  <img src=".figures/aml-dashboard_batch_prediction.png" width="750">
</div>

### Real-Time Prediction Tab

Experience real-time image prediction, utilizing either an AML Model or a Neural Network.

To predict the output of webcam images in real-time using an AML Model, follow these steps:

1. Toggle the `prediction` switch in the *Predict for AML* section and activate the `video` switch in the *webcam* section.
2. The predictions will be displayed in the *Results AML* plot.

> * Ensure that you have access to at least one AML Model to facilitate the real-time prediction process.

Reviewer: Put before step 1

<div align="center">
  <img src=".figures/aml-dashboard_real_time.png" width="750">
</div>

### Context Broker Tab

Effortlessly manage the Context Broker, allowing you to create and update data to an entity and get the solution of this data.

Here's how to use it:

1. Specify the Fiware Node parameters and Context Broker entity ID and attributes.
2. Click on the `Create` button to create the node using the provided parameters.
3. The *Fiware Node Status* will update to indicate whether the node has been created successfully.
3. Drag and drop an image or upload a file in the designated area to upload data.
4. Click the `Post Data` button to send the data to the Context Broker.
5. The *Data Status* will update to indicate whether the data has been successfully posted.
6. Once the solution is received, it will be displayed in the *Context Broker Solution* section and the *Solution Status* will update to **Solution received !** to indicate successful retrieval.

> * Make sure that the Context Broker is running to ensure successful interaction and data exchange.

Reviewer: Put before step 1

```

### Building from sources

If **AML-IP** is not already installed in your system, follow one of the methods outlined in its [documentation site](https://aml-ip.readthedocs.io/en/latest/rst/developer_manual/installation/sources/linux/linux.html).

### Requirements

Before proceeding further, ensure that you have all requirements installed in your system. 
Firstly, some Python dependencies are required. This version of the AML-Dashboard uses `Python 3.11`, installed in a conda virtual environment to install such dependencies, thus avoiding polluting the user's installation. Create a virtual environment and install Python dependencies.

```bash
curl https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -o Miniconda3-latest-Linux-x86_64.sh

#To activate the conda environment automatically when opening shell. For changes to take effect, close and re-open current shell.
bash Miniconda3-latest-Linux-x86_64.sh

#To activate the conda environment manually each time you open the shell:
eval "$(/home/user/miniconda3/bin/conda shell.zsh hook)" 

conda create --name aml-ip-venv python=3.11
conda activate aml-ip-venv
cd AML-Dashboard
pip3 install -r requirements.txt

> ```

Author: Same https://github.com/eProsima-Private/AML-Tools/pull/12#discussion_r1744851309

Reviewer: Done

# AML-Dashboard

## Installation

To get started with AML-Dashboard, follow these steps.

Clone this repository using the following command:

```bash
git clone https://github.com/eProsima-Private/AML-Dashboard.git
```

### Building from sources

If **AML-IP** is not already installed in your system, follow one of the methods outlined in its [documentation site](https://aml-ip.readthedocs.io/en/latest/rst/developer_manual/installation/sources/linux/linux.html).

### Requirements

Before proceeding further, ensure that you have all requirements installed in your system. 

> Firstly, some Python3 dependencies are required. This version of the AML-Dashboard uses `Python3.11`, installed in a virtual environment to install such dependencies, thus avoiding polluting the user's installation. 

Author: ```suggestion
Firstly, some Python dependencies are required. This version of the AML-Dashboard uses `Python3.11`, installed in a virtual environment to install such dependencies, thus avoiding polluting the user's installation. 
```

Reviewer: Done

> Create a virtual environment and install python3 dependencies.

Author: ```suggestion
Create a virtual environment and install Python dependencies.
```

		```

	To stop an inference node, simply run `ctrl+C`.

### Frontend Marcelle Framework

To run the frontend Marcelle framework, follow the steps below:

**Launch the dashboard**

1. Navigate to the [*frontend/aml_dashboard*](frontend/aml_dashboard) directory

	```bash
	cd frontend/aml_dashboard
	```

2. Run the dashboard

	```bash
	npm run dev
	```

With these steps completed, you're all set to utilize the AML-Dashboard, seamlessly integrating backend server functionalities with the Marcelle framework frontend.

## Usage

With the AML-Dashboard up and running, you can access it at the following address:

```bash
http://localhost:5173/
```

### Data Management Tab

Efficiently manage and create new datasets.

To create a dataset, simply follow these steps:

1. Activate the `video` switch in the *webcam* section to enable webcam capture.
2. Specify the label of the dataset in the *Instance label* section.
3. Click on the `Hold to record instances` button in the *Capture instances to the training set* to start recording instances.
4. Once recorded, the dataset will be promptly displayed in the *dataset browser* section for easy access and management.

<div align="center">

>   <img src=".figures/aml-dashboard_data_management.png" width="750">

Author: I think you have uploaded the incorrect image. The one provided is from the Context Broker Tab, not the Data Management Tab.

Reviewer: Done, sorry for the confusion

Author: No problem!

# AML-Dashboard

## Installation

To get started with AML-Dashboard, follow these steps.

Clone this repository using the following command:

```bash
git clone https://github.com/eProsima-Private/AML-Dashboard.git
```

### Building from sources

If **AML-IP** is not already installed in your system, follow one of the methods outlined in its [documentation site](https://aml-ip.readthedocs.io/en/latest/rst/developer_manual/installation/sources/linux/linux.html).

### Requirements

Before proceeding further, ensure that you have all requirements installed in your system. 

> Firstly, some python3 dependencies are required. This version of the AML-Dashboard uses python3.11, installed in a conda virtual environment to install such dependencies, thus avoiding polluting the user's installation. Create a virtual environment and install python3 dependencies.

Author: ```suggestion
Firstly, some Python dependencies are required. This version of the AML-Dashboard uses `Python 3.11`, installed in a conda virtual environment to install such dependencies, thus avoiding polluting the user's installation. Create a virtual environment and install Python dependencies.
```

Reviewer: Done

2. Specify the label of the dataset in the *Instance label* section.
3. Click on the `Hold to record instances` button in the *Capture instances to the training set* to start recording instances.
4. Once recorded, the dataset will be promptly displayed in the *dataset browser* section for easy access and management.

<div align="center">
  <img src=".figures/aml-dashboard_data_management.png" width="750">
</div>

### Training Tab

Effortlessly train models using either AML or Neural Network.

To train a model using AML, follow these steps:

* Note: Before starting, please ensure that at least one **Computing Node** is running to facilitate the training process.

1. Specify the number of parallel trainings (executions) you wish to run.
2. Define the number of iterations per execution.
3. Set the percentage of dataset to distribute in each execution.
4. Click on the `Train` button in the *AML Training Launcher* to initiate the training process.
5. Once the training is completed, the model status will appear as **Finished :)** in the *AML Status*.

<div align="center">
  <img src=".figures/aml-dashboard_training.png" width="750">

> </div>

Author: ```suggestion
Effortlessly train models using either AML or Neural Network.

* Note: Before starting, please ensure that at least one **Computing Node** is running to facilitate the training process.

To train a model using AML, follow these steps:

1. Specify the number of parallel trainings (executions) you wish to run.
2. Define the number of iterations per execution.
3. Set the percentage of dataset to distribute in each execution.
4. Click on the `Train` button in the *AML Training Launcher* to initiate the training process.
5. Once the training is completed, the model status will appear as **Finished :)** in the *AML Status*.

<div align="center">
  <img src=".figures/aml-dashboard_training.png" width="750">
</div>
```

### Fetching Tab

Retrieve trained AML Models effortlessly.

To fetch a model, follow these steps:

* Note: make sure that at least one **Model Manager Sender Node** is running to facilitate the model fetching process.

1. Click on the `Search for statistics` button in the *AML Statistics Fetcher*.
2. Once the statistics are received, the status will appear as **Statistics received !** in the *AML Collaborative Learning Status*.
3. Click on the `Request model` button in the *AML Model Fetcher*.
4. Once the model is received, the status will change to **Model received !** in the *AML Collaborative Learning Status*.

<div align="center">
  <img src=".figures/aml-dashboard_fetching.png" width="750">

> </div>

Author: ```suggestion
Retrieve trained AML Models effortlessly.

* Note: make sure that at least one **Model Manager Sender Node** is running to facilitate the model fetching process.

To fetch a model, follow these steps:

1. Click on the `Search for statistics` button in the *AML Statistics Fetcher*.
2. Once the statistics are received, the status will appear as **Statistics received !** in the *AML Collaborative Learning Status*.
3. Click on the `Request model` button in the *AML Model Fetcher*.
4. Once the model is received, the status will change to **Model received !** in the *AML Collaborative Learning Status*.

<div align="center">
  <img src=".figures/aml-dashboard_fetching.png" width="750">
</div>
```

### Batch Prediction Tab

Effortlessly predict the output of a batch of inputs using an AML Model.

To predict the output of a dataset, follow these steps:

* Note: Before staring, ensure that you have access to at least one AML Model to facilitate the batch prediction process.

1. Click on the `Update predictions` button in the *Algebraic Machine Learning*.
2. The predictions will be displayed in the *Results Algebraic Machine Learning* plot.

<div align="center">
  <img src=".figures/aml-dashboard_batch_prediction.png" width="750">

> </div>

Author: ```suggestion
### Batch Prediction Tab

Effortlessly predict the output of a batch of inputs using an AML Model.

* Note: Before staring, ensure that you have access to at least one AML Model to facilitate the batch prediction process.

To predict the output of a dataset, follow these steps:

1. Click on the `Update predictions` button in the *Algebraic Machine Learning*.
2. The predictions will be displayed in the *Results Algebraic Machine Learning* plot.

<div align="center">
  <img src=".figures/aml-dashboard_batch_prediction.png" width="750">
</div>
```

### Real-Time Prediction Tab

Experience real-time image prediction, utilizing either an AML Model or a Neural Network.

To predict the output of webcam images in real-time using an AML Model, follow these steps:

* Note: Before starting, ensure that you have access to at least one AML Model to facilitate the real-time prediction process.

1. Toggle the `prediction` switch in the *Predict for AML* section and activate the `video` switch in the *webcam* section.
2. The predictions will be displayed in the *Results AML* plot.

<div align="center">
  <img src=".figures/aml-dashboard_real_time.png" width="750">

> </div>

Author: ```suggestion
### Real-Time Prediction Tab

Experience real-time image prediction, utilizing either an AML Model or a Neural Network.

* Note: Before starting, ensure that you have access to at least one AML Model to facilitate the real-time prediction process.

To predict the output of webcam images in real-time using an AML Model, follow these steps:

1. Toggle the `prediction` switch in the *Predict for AML* section and activate the `video` switch in the *webcam* section.
2. The predictions will be displayed in the *Results AML* plot.

<div align="center">
  <img src=".figures/aml-dashboard_real_time.png" width="750">
</div>
```

### Context Broker Tab

Effortlessly manage the Context Broker, allowing you to create and update data to an entity and get the solution of this data.

Here's how to use it:

* Note: Before starting, make sure that the Context Broker is running to ensure successful interaction and data exchange.

1. Specify the Fiware Node parameters and Context Broker entity ID and attributes.
2. Click on the `Create` button to create the node using the provided parameters.
3. The *Fiware Node Status* will update to indicate whether the node has been created successfully.
3. Drag and drop an image or upload a file in the designated area to upload data.
4. Click the `Post Data` button to send the data to the Context Broker.
5. The *Data Status* will update to indicate whether the data has been successfully posted.
6. Once the solution is received, it will be displayed in the *Context Broker Solution* section and the *Solution Status* will update to **Solution received !** to indicate successful retrieval.

<div align="center">
  <img src=".figures/aml-dashboard_context_broker.png" width="750">

> </div>

Author: ```suggestion
### Context Broker Tab

Effortlessly manage the Context Broker, allowing you to create and update data to an entity and get the solution of this data.

* Note: Before starting, make sure that the Context Broker is running to ensure successful interaction and data exchange.

Here's how to use it:

1. Specify the Fiware Node parameters and Context Broker entity ID and attributes.
2. Click on the `Create` button to create the node using the provided parameters.
3. The *Fiware Node Status* will update to indicate whether the node has been created successfully.
3. Drag and drop an image or upload a file in the designated area to upload data.
4. Click the `Post Data` button to send the data to the Context Broker.
5. The *Data Status* will update to indicate whether the data has been successfully posted.
6. Once the solution is received, it will be displayed in the *Context Broker Solution* section and the *Solution Status* will update to **Solution received !** to indicate successful retrieval.

<div align="center">
  <img src=".figures/aml-dashboard_context_broker.png" width="750">
</div>
```
